File: PresentationStateContainers/Android_MVP/app/src/main/java/com/soxjke/android_mvp/LoginPresenter.kt
package com.soxjke.android_mvp

import android.util.Log

class LoginPresenter(private val fragment: LoginFragment) {
    private var login = ValidatableField<LoginValidator.Error>()
    private var loginValidator = LoginValidator()
    private var password = ValidatableField<PasswordValidator.Error>()
    private var passwordValidator = PasswordValidator()

    init {
        present()
    }

    fun present() {
        fragment.render(makeProps())
    }

    private fun makeProps(): LoginFragment.Props {
        return LoginFragment.Props(
            makeLoginField(),
            makePasswordField(),
            loginValidator.validate(login.text) == null && passwordValidator.validate(password.text) == null,
            this::loginAction
        )
    }

    private fun makeLoginField(): ValidatableField<LoginValidator.Error> {
        return ValidatableField<LoginValidator.Error>(
            login.error,
            login.text,
            this::loginStartEditing,
            this::loginEditingChanged,
            this::loginEndEditing
        )
    }

    private fun loginStartEditing(text: String) {
        login.error = null
        present()
    }

    private fun loginEditingChanged(text: String) {
        login.text = text
        present()
    }

    private fun loginEndEditing(text: String) {
        login.error = loginValidator.validate(text);
        present()
    }

    private fun makePasswordField(): ValidatableField<PasswordValidator.Error> {
        return ValidatableField<PasswordValidator.Error>(
            password.error,
            password.text,
            this::passwordStartEditing,
            this::passwordEditingChanged,
            this::passwordEndEditing
        )
    }

    private fun passwordStartEditing(text: String) {
        password.error = null
        present()
    }

    private fun passwordEditingChanged(text: String) {
        password.text = text
        present()
    }

    private fun passwordEndEditing(text: String) {
        password.error = passwordValidator.validate(text);
        present()
    }

    private fun loginAction() {
        // Call login method
        Log.d("","LoginPresenter now will call login with:\n${login.text}\n${password.text}")
    }
}

File: PresentationStateContainers/iOS_MVP/iOS_MVP/Screens/Login/LoginPresenter.swift
//
//  LoginPresenter.swift
//  iOS_MVP
//
//  Created by Petro Korienev on 5/14/20.
//  Copyright © 2020 Petro Korienev. All rights reserved.
//

import Foundation

class LoginPresenter {
    typealias Props = LoginViewController.Props
    weak var controller: LoginViewController?
    
    private var login: LoginValidatableTextField.Props = .defaultValue
    private var loginValidator = LoginValidator()
    private var password: PasswordValidatableTextField.Props = .defaultValue
    private var passwordValidator = PasswordValidator()
    
    init(controller: LoginViewController) {
        self.controller = controller
        present()
    }
    
    func present() {
        controller?.render(props: makeProps())
    }
    
    private func makeProps() -> Props {
        return Props(
            login: makeLoginField(),
            password: makePasswordField(),
            action: makeAction()
        )
    }
    
    private func makeLoginField() -> LoginValidatableTextField.Props {
        return .init(error: login.error,
                     text: login.text,
                     startEditing: weakify(self, LoginPresenter.loginStartEditing),
                     changedEditing: weakify(self, LoginPresenter.loginEditingChanged),
                     endEditing: weakify(self, LoginPresenter.loginEndEditing)
                )
    }
    
    private func loginStartEditing(_ text: String) {
        login.error = nil
        present()
    }
    
    private func loginEditingChanged(_ text: String) {
        login.text = text
        present()
    }
    
    private func loginEndEditing(_ text: String) {
        login.error = loginValidator.validate(string: text);
        present()
    }
    
    private func makePasswordField() -> PasswordValidatableTextField.Props {
        return .init(error: password.error,
                     text: password.text,
                     startEditing: weakify(self, LoginPresenter.passwordStartEditing),
                     changedEditing: weakify(self, LoginPresenter.passwordEditingChanged),
                     endEditing: weakify(self, LoginPresenter.passwordEndEditing)
                )
    }
    
    private func passwordStartEditing(_ text: String) {
        password.error = nil
        present()
    }
    
    private func passwordEditingChanged(_ text: String) {
        password.text = text
        present()
    }
    
    private func passwordEndEditing(_ text: String) {
        password.error = passwordValidator.validate(string: text);
        present()
    }
    
    private func makeAction() -> ActionableButton.Props {
        return .init(isEnabled: loginValidator.validate(string: login.text) == nil &&
                                passwordValidator.validate(string: password.text) == nil,
                     action: weakify(self, LoginPresenter.loginAction))
    }
            
    private func loginAction() {
        // Call login method
        print("LoginPresenter now will call login with:\n\(login.text)\n\(password.text)")
    }
}
